Dev imu interface

ThorsHamster opened the pull request:
Closes #176 

Adds public interface for Inertial Measurement Unit. Includes abstract and concrete classes. See wiki for UML class diagram.

ThorsHamster committed:
Add of file/folder structure for imu and intial failed test

ThorsHamster committed:
Add of sensitivity types for Gyroscope and Accelerometer

ThorsHamster committed:
Add of stub class for i2c

ThorsHamster committed:
Add of imu interface with unittest draft

ThorsHamster committed:
Add of CMakeLists.txt for i2c and imu

ThorsHamster committed:
Added mock for i2c

ThorsHamster committed:
Add of mock libraries for i2c_handler and generic_imu

ThorsHamster committed:
Add of GetGyroscopeSensitivity in Interface

ThorsHamster committed:
Add of unit tests for GetGyroscopeSensitivity

ThorsHamster committed:
Add of unit test for default Gyroscope sensitivity value

ThorsHamster committed:
Add of unit tests for accelerometer get and set of sensitivity

ThorsHamster committed:
Add of Interface for GetAccelerometerSensitivity

ThorsHamster committed:
Add of unit tests for GetGyroscope, GetAccelerometer, GetMagnetometer, GetTemperature

not buildable yet

ThorsHamster committed:
Moved implementation to InertialMeasurement

ThorsHamster committed:
Merge branch 'dev' into dev_imu_interface

ThorsHamster committed:
Reactivation of unit tests

ThorsHamster committed:
add of standard return values

ThorsHamster committed:
minor beautifying. not linkable atm.

ThorsHamster committed:
Fix of linking error

All methods have to be pure virtual (=0). Thanks to @elheck.

ThorsHamster committed:
Merge branch 'dev' into dev_imu_interface

ThorsHamster committed:
Removal of unnecessary include

ThorsHamster committed:
Moved mock_library to folder mock_libraries

ThorsHamster committed:
restructuring of tests/CMakeLists.txt

ThorsHamster committed:
Rename of imu_tests.cpp to imu_interface_tests.cpp

ThorsHamster committed:
Add of mock libraries for generic_imu and mpu9250

ThorsHamster committed:
Add of src of generic imu and mpu9250

ThorsHamster committed:
Removal of mock library for mpu9250

ThorsHamster committed:
Renaming of mock library generic_imu to mpu9250

ThorsHamster committed:
Add of empty line for better readability

ThorsHamster committed:
Add of editorconfig config, for brackets on same line

for Google C++ StyleGuide

ThorsHamster committed:
fixing of new path

ThorsHamster committed:
Add of mpu9250.cpp and restructuring

ThorsHamster committed:
Implementation of intertial_measurement.cpp

ThorsHamster committed:
Add of stub values

ThorsHamster committed:
Adapted StyleGuide

ThorsHamster committed:
Adapted StyleGuide

ThorsHamster committed:
Renaming of protected member

ThorsHamster committed:
Add of set of protected member

ThorsHamster committed:
change of protected members to private

ThorsHamster committed:
removal of mock_library mpu9250.hpp

ThorsHamster committed:
Add of private members

ThorsHamster committed:
Renaming of protected attribute mpu9250

ThorsHamster committed:
Fix of wrong define

ThorsHamster committed:
Removal of not needed mock i2c_handler.hpp. Use of "original" one

ThorsHamster committed:
Replacement of type i2c::I2CHandler with std::unique_ptr<i2c::I2CHandler>

ThorsHamster committed:
Removal of unnecessary attribute

ThorsHamster committed:
Removal of redundant attributes in parent and child class

ThorsHamster committed:
Also needed to be renamed

ThorsHamster committed:
Refactor of unit test to use googletests ::testing::Test with SetUp function

ThorsHamster committed:
Removal of i2c read method

ThorsHamster committed:
Adapted architecture

Like discussed in Jourfixe

ThorsHamster committed:
StyleGuide modifications

ThorsHamster committed:
Changed to trailing return type

ThorsHamster committed:
Changed to trailing return type

ThorsHamster committed:
Add of doxygen docu

ThorsHamster committed:
Minor changes of doxygen docu

ThorsHamster committed:
Minor changes of doxygen docu

ThorsHamster committed:
Merge branch 'dev' into dev_imu_interface

ThorsHamster committed:
Removal of src files. Shifted to mock libraries for first pull request of interface

ThorsHamster committed:
Shifted i2c to mock_libraries.

ThorsHamster committed:
Removal of CodeFactor issue

elheck commented on squidDrone/src/imu/imu_interface.hpp:
```diff
@@ -0,0 +1,87 @@
+      /// @brief  The custom constructor is the one to be used. 
+      /// @param  i2c_handler Unique pointer to I2C Handler defined by hal driver 
+      ///
+      explicit InertialMeasurementInterface(std::unique_ptr<i2c::I2CHandler> i2c_handler): mpu9250_(std::make_unique<imu::Mpu9250>(std::move(i2c_handler))){};
```
Why does the IMU interface have a concrete inertial measurement unit as a protected variable.
It should be a GenericInertialMeasurementUnit, otherwise you would have to change this if we would use another concrete inertial measurement unit.
The question your should be asking yourself should be: What would you need to do, in order to switch to another concrete IMU. If the answer is, making changes to the interface, than this might indicate an error.
An interface should not be responsible for initializing a concrete object in constructor.
You could add a GenericIMU object as a parameter for the constructor and the constructor just moves it to its own protected variable. The concrete InertialMeasurement class would then need to construct a concrete object and give it to the parent class constructor in the initializer list.
This would remove the decision about the concrete InertialMeasrementUnit from the interface and move it to the concrete implementation. This would mean, if we want to change the conrete InertialMeasurementUnit, we would need to make changes to the concrete implementation of the InertialMeasurementInterface and this is much better than changing the interface.

ThorsHamster replied:
Done like suggested

elheck commented on squidDrone/src/imu/inertial_measurement.hpp:
```diff
@@ -0,0 +1,35 @@
+
+  ///
+  /// @class InertialMeasurement 
+  /// @brief The InertialMeasurement provides an interface for the Inertial Measurement Unit.
```
No it doesn't, it is the concrete implementation of the interface

ThorsHamster replied:
Fixed

elheck commented on squidDrone/src/types/accelerometer_sensitivity.hpp:
```diff
@@ -0,0 +1,22 @@
+  ///
+  /// @brief A enum for Accelerometer Sensitivity of Inertial Measurement Unit
+  ///
+  enum class AccelerometerSensitivity: int{
```
Wouldn't one sensitivity types file be enough? I think it will bloat the types folder, what do you think?

ThorsHamster replied:
We can discuss about this. If we change the IMU some day and the new IMU has only 3 sensitivity possibilities for accelerometer, but not for gyroscope, than we would have to change then. Atm. we can also combine both enum classes. How should we name them then?

elheck replied:
Ah I see your point. I can see the benefit of keeping them seperate.

elheck commented on squidDrone/tests/imu/imu_interface_tests.cpp:
```diff
@@ -0,0 +1,120 @@
+  };
+
+  TEST_F(ImuInterfaceTests, interface_set_gyroscope_sensitivity_finest){
+    types::GyroscopeSensitivity sensitivity = types::GyroscopeSensitivity::FINEST;
```
Please apply almost always auto.
The type is clearly defined by types::GyroscopeSensitivity::FINEST, so we do not need the explicit type on the left side.
Same with the other tests.

ThorsHamster replied:
Fixed

elheck commented on squidDrone/src/imu/imu_interface.hpp:
```diff
@@ -0,0 +1,87 @@
+  /// @class InertialMeasurementInterface 
+  /// @brief The InertialMeasurementInterface provides an interface for the Inertial Measurement Unit.
+  ///
+  class InertialMeasurementInterface {
```
Can setting a sensitivity, fail? If yes, than these cases have not been handled, because there is no error handling.

ThorsHamster replied:
I think it is not really possible to fail. Otherwise I will add this later.

elheck replied:
Ok

elheck commented on squidDrone/src/imu/inertial_measurement.hpp:
```diff
@@ -0,0 +1,35 @@
+      auto GetMagnetometer(void) noexcept -> types::EuclideanVector<float> override;
+      auto GetTemperature(void) noexcept -> int override;
+  };
+
```
In the class diagram there is a private MPU9250 member in this class.

ThorsHamster replied:
UML Model updated

elheck reviewed (changes requested):
Please merge the current dev branch into your branch, to resolve branch conflict in CMakeLists.txt

ThorsHamster committed:
Merge branch 'dev' into dev_imu_interface

ThorsHamster committed:
Add of missing abstract_motor

ThorsHamster committed:
Removal of concrete imu from imu interface

Like suggested in review, moved initialization of concrete interface to implementation, not interface.

ThorsHamster committed:
Fix of class doxygen header

ThorsHamster committed:
change variable types to almost always auto

ThorsHamster committed:
Rename from generic_imu_ to imu_

elheck commented on squidDrone/src/imu/inertial_measurement.hpp:
```diff
@@ -0,0 +1,35 @@
+#ifndef SRC_IM_HPP_
```
SRC_IM_HPP is not very unique. Please choose a more unique keyword to prevent inclusion conflicts

ThorsHamster replied:
Renaming done

ThorsHamster committed:
Renaming of include guard

elheck reviewed (approved):
LGTM
